client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala

```diff
@@ -79,14 +81,17 @@ class ChangePartitionManager(
                 batchHandleChangePartitionExecutors.submit {
                   new Runnable {
                     override def run(): Unit = {
-                      val distinctPartitions = requests.synchronized {
-                        // For each partition only need handle one request
-                        requests.asScala.filter { case (partitionId, _) =>
-                          !inBatchPartitions.get(shuffleId).contains(partitionId)
-                        }.map { case (partitionId, request) =>
-                          inBatchPartitions.get(shuffleId).add(partitionId)
-                          request.asScala.toArray.maxBy(_.epoch)
-                        }.toArray
+                      val distinctPartitions = {
+                        requests.asScala.map { case (partitionId, request) =>
+                          locks(partitionId % locks.length).synchronized {
+                            if (!inBatchPartitions.contains(partitionId)) {
+                              inBatchPartitions.get(shuffleId).add(partitionId)
+                              Some(request.asScala.toArray.maxBy(_.epoch))
+                            } else {
+                              None
+                            }
+                          }
+                        }.filter(_.isDefined).map(_.get).toArray
```

Comment:
I am not so sure of this change - it feels like it will have a lot more contention - as each entry in the `requests` map will need to acquire a lock - while uncontented locks are cheap to acquire, it is not zero cost still.

If you have a test bed to validate perf, how about this ?
```
                      val batchPartitions = inBatchPartitions.get(shuffleId)
                      val distinctPartitions = requests.synchronized {
                        // For each partition only need handle one request
                        requests.asScala.filter { case (partitionId, _) =>
                          !batchPartitions.contains(partitionId)
                        }.map { case (partitionId, request) =>
                          batchPartitions.add(partitionId)
                          request.asScala.maxBy(_.epoch)
                        }.toArray
                      }
```

Reply:
based on our observation in our production system, it won't bring more competition ... 

if we use `requests.synchronized`, all celeborn-dispatcher threads + all celeborn-client-life-cycle-manager-change-partition-executor will all compete for the same object for locking, even they are likely working on different partitions ... check the following screenshots


![image](https://github.com/apache/celeborn/assets/678008/37fe018f-5084-42a7-b727-91f03b9d6913)

![image](https://github.com/apache/celeborn/assets/678008/8a17f576-1d20-4d17-956b-977a2ad73db0)

after this change, with a huge spark application of 300TB shuffle data, I don't see such intensive locking competition anymore

Reply:
Agree, in my analysis, I was thinking of locking on `batchPartitions ` and not `requests` ... but I messed up the diffs here (and below).

But hopefully you get the idea.

Reply:
>  it feels like it will have a lot more contention - as each entry in the requests map will need to acquire a lock

To reduce the frequency of acquiring locks, I think we can calculate the lock buckets for each partition ids first, then group the partition ids by the lock bucket,  then acquire lock and process each group (in random order). Though I'm not sure how beneficial this will be.